src/util/datablock/datablock.c

```diff
@@ -72,93 +73,132 @@ void _DataBlock_AddBlocks(DataBlock *dataBlock, size_t blockCount) {
     dataBlock->itemCap = dataBlock->blockCount * BLOCK_CAP;
 }
 
+void static inline _DataBlock_MarkItemAsDeleted(const DataBlock *dataBlock, char* item) {
+    for(int i = 0; i < dataBlock->itemSize; i++) item[i] = DELETED_MARKER;
+}
+
+void static inline _DataBlock_MarkItemAsUndelete(const DataBlock *dataBlock, char* item) {
+    item[0] = !DELETED_MARKER;
+}
+
+int static inline _DataBlock_IsItemDeleted(const DataBlock *dataBlock, char* item) {
+    char deleted = 0xFF;
+    for(int i = 0; i < dataBlock->itemSize; i++) deleted &= (item[i] & DELETED_MARKER);
+    return deleted;
+}
+
+// Checks to see if idx is within global array bounds
+// array bounds are between 0 and itemCount + #deleted indices
+// e.g. [3, 7, 2, D, 1, D, 5] where itemCount = 5 and #deleted indices is 2
+// and so it is valid to query the array with idx 6.
+int static inline _DataBlock_IndexOutOfBounds(const DataBlock *dataBlock, size_t idx) {
+    return (idx < 0 || idx >= (dataBlock->itemCount + array_len(dataBlock->deletedIdx)));
```

Comment:
size_t is unsigned, so first condition is impossible.

Reply:
👍